File: temp/rsa_crypt.py
import math

# Helper function: Compute the greatest common divisor
def gcd(a, b):
	while b != 0:
		a, b = b, a % b
	return a

# Helper function: Compute the modular multiplicative inverse
def mod_inverse(e, phi):
	t, new_t = 0, 1
	r, new_r = phi, e

	while new_r != 0:
		quotient = r // new_r
		t, new_t = new_t, t - quotient * new_t
		r, new_r = new_r, r - quotient * new_r

	if r > 1:
		raise ValueError("e is not invertible")
	if t < 0:
		t += phi

	return t

# Helper function: Compute the power modulo
def mod_pow(base, exp, mod):
	result = 1
	base = base % mod
	while exp > 0:
		if exp % 2 == 1:
			result = (result * base) % mod
		exp = exp // 2
		base = (base * base) % mod
	return result

# Function: Generate RSA keys
def generate_keys(p, q):
	n = p * q
	phi = (p - 1) * (q - 1)
	e = 3

	# Ensure e and phi are coprime
	while gcd(e, phi) != 1:
		e += 2

	d = mod_inverse(e, phi)

	return {'publicKey': {'e': e, 'n': n}, 'privateKey': {'d': d, 'n': n}}

# Function: Encrypt message chunks using RSA encryption
def encrypt_string_chunks(public_key, message, chunk_size):
	e, n = public_key['e'], public_key['n']
	cipher_text = []

	for i in range(0, len(message), chunk_size):
		chunk = message[i:i+chunk_size]
		encrypted_chunk = []

		for char in chunk:
			m = ord(char)
			encrypted_chunk.append(mod_pow(m, e, n))

		cipher_text.append(encrypted_chunk)

	return cipher_text

# Function: Decrypt message chunks using RSA decryption
def decrypt_string_chunks(private_key, cipher_text):
	d, n = private_key['d'], private_key['n']
	message = []

	for encrypted_chunk in cipher_text:
		decrypted_chunk = []

		for c in encrypted_chunk:
			m = mod_pow(c, d, n)
			decrypted_chunk.append(chr(m))

		message.append(''.join(decrypted_chunk))

	return ''.join(message)

# Example usage
p = 61  # Choose two prime numbers
q = 53
chunk_size = 4  # Example chunk size

keys = generate_keys(p, q)
public_key = keys['publicKey']
private_key = keys['privateKey']

message = "Hello, Roblox!"
encrypted_chunks = encrypt_string_chunks(public_key, message, chunk_size)
decrypted_message = decrypt_string_chunks(private_key, encrypted_chunks)

print("Original Message:", message)
print("Encrypted Chunks:")
for i, chunk in enumerate(encrypted_chunks):
	print(i, ' '.join(map(str, chunk)))
print("Decrypted Message:", decrypted_message)

File: temp/rsa_crypt.lua
local RSA = {}

-- Helper function: Compute the greatest common divisor
local function gcd(a, b)
	while b ~= 0 do
		a, b = b, a % b
	end
	return a
end

-- Helper function: Compute the modular multiplicative inverse
local function modInverse(e, phi)
	local t, newT = 0, 1
	local r, newR = phi, e
	while newR ~= 0 do
		local quotient = math.floor(r / newR)
		t, newT = newT, t - quotient * newT
		r, newR = newR, r - quotient * newR
	end
	if r > 1 then
		error("e is not invertible")
	end
	if t < 0 then
		t = t + phi
	end
	return t
end

-- Helper function: Compute the power modulo
local function modPow(base, exp, mod)
	local result = 1
	base = base % mod
	while exp > 0 do
		if exp % 2 == 1 then
			result = (result * base) % mod
		end
		exp = math.floor(exp / 2)
		base = (base * base) % mod
	end
	return result
end

-- Function: Generate RSA keys
function RSA.generateKeys(p, q)
	local n = p * q
	local phi = (p - 1) * (q - 1)
	local e = 3

	-- Ensure e and phi are coprime
	while gcd(e, phi) ~= 1 do
		e = e + 2
	end

	local d = modInverse(e, phi)

	return {publicKey = {e = e, n = n}, privateKey = {d = d, n = n}}
end

-- Function: Encrypt message chunks using RSA encryption
function RSA.encryptStringChunks(publicKey, message, chunkSize)
	local e, n = publicKey.e, publicKey.n
	local cipherText = {}
	for i = 1, #message, chunkSize do
		local chunk = message:sub(i, i + chunkSize - 1)
		local encryptedChunk = {}
		for j = 1, #chunk do
			local m = string.byte(chunk, j)
			table.insert(encryptedChunk, modPow(m, e, n))
		end
		table.insert(cipherText, encryptedChunk)
	end
	return cipherText
end

-- Function: Decrypt message chunks using RSA decryption
function RSA.decryptStringChunks(privateKey, cipherText)
	local d, n = privateKey.d, privateKey.n
	local message = {}
	for i = 1, #cipherText do
		local encryptedChunk = cipherText[i]
		local decryptedChunk = {}
		for j = 1, #encryptedChunk do
			local c = encryptedChunk[j]
			local m = modPow(c, d, n)
			table.insert(decryptedChunk, string.char(m))
		end
		table.insert(message, table.concat(decryptedChunk))
	end
	return table.concat(message)
end

-- Example usage
local p = 61  -- Choose two prime numbers
local q = 53
local chunkSize = 4  -- Example chunk size

local keys = RSA.generateKeys(p, q)
local publicKey = keys.publicKey
local privateKey = keys.privateKey

local message = "Hello, Roblox!"
local encryptedChunks = RSA.encryptStringChunks(publicKey, message, chunkSize)
local decryptedMessage = RSA.decryptStringChunks(privateKey, encryptedChunks)

print("Original Message:", message)
print("Encrypted Chunks:")
for i, chunk in ipairs(encryptedChunks) do
	print(i, table.concat(chunk, " "))
end
print("Decrypted Message:", decryptedMessage)
